staging/src/k8s.io/apiserver/pkg/storage/etcd3/featuresupportchecker/feature_support_checker.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package featuresupportchecker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/storage"
+)
+
+var (
+	// This is a error type that we used in checking RequestWatchProgress support
+	errNoSupportRequestWatchProgress = errors.New("requestWatchProgress is not supported for this cluster")
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_25 = version.MustParseSemantic("3.4.25")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_8  = version.MustParseSemantic("3.5.8")
+	// featureSupportCache store result of if feature is supported or not.
+	featureSupportCache sync.Map
```

Comment:
I'm a little bit worry about this if it is a good practice to use it here or not, but i think this type of solution will solve our problem to check with storage interface and cache it here and use this cache later without knowing about storage.

Reply:
My recommendation would be to not use global `sync.Map`, but to move `DefaultFeatureSupportChecker` from a storage field, to a global variable. 

`DefaultFeatureSupportChecker` would implement two methods private `endpointSupports(ctx, client, endpoint, feature)` that is called by storage and results are stored in per endpoint cache and a `Supports(feature)` that returns based on previously cached values.

Reply:
Thanks, Ok i'll do that

Reply:
I have one thing to worry about, if i define `featureSupportCache` inside of `DefaultFeatureSupportChecker`, we should force it to only constructed once to access the cache everywhere. Am i right?

Reply:
Right, see how `utilfeature.DefaultFeatureGate` is defined.

Reply:
Done